proposals/NNNN-consistent-naming-for-dx-intrinsics.md

```diff
@@ -0,0 +1,224 @@
+  `llvm.dx.wave.getlaneindex`, not `llvm.dx.wave.is.first.lane`
+- Namespace handle ops under "resource": `llvm.dx.resource.frombinding`,
+  `llvm.dx.resource.fromheap`. Include `llvm.dx.resource.pointer` or
+  `llvm.dx.resource.access` here for getting a pointer in a namespace for
+  resources.
+- Namespace other resource ops with the verb and then specialize with the type:
+  `llvm.dx.store.typedbuffer`, `llvm.dx.load.texture`, `llvm.dx.sample`,
+  `llvm.dx.sample.bias`, `llvm.dx.getdimensions.typedbuffer`, `llvm.dx.gather`,
+  `llvm.dx.getsampleposition.Texture2DMS`
+- Use other namespaces where it makes sense, like `wave`, `quad`, and
+  `rayquery`.
+
+## Detailed design
+
+Applying these naming conventions to a number of ops, we can get a taste for
+how the naming convention will work. What follows are a large number of
+examples and some open questions around a few of them.
+
+#### Thread and Group IDs
+```llvm
+llvm.dx.threadid ; (ThreadId)
+llvm.dx.groupid ; (GroupId)
+llvm.dx.threadidingroup ; (ThreadIdInGroup)
+llvm.dx.flattenedthreadidingroup ; (FlattenedThreadIdInGroup)
+```
+
+Open questions:
+- Is "flattenedthreadidingroup" too unwieldy?
+
+#### Resource ops
+```llvm
+llvm.dx.resource.handlefrombinding ; (CreateHandleFromBinding)
+llvm.dx.resource.handlefromheap ; (CreateHandleFromHeap)
+llvm.dx.resource.annotatehandle ; (AnnotateHandle)
+llvm.dx.resource.load.rawbuffer ; (RawBufferLoad)
+llvm.dx.resource.load.typedbuffer ; (BufferLoad)
+llvm.dx.resource.load.texture ; (TextureLoad)
+llvm.dx.resource.load.cbufferrow ; (CBufferLoadLegacy)
+llvm.dx.resource.load.cbuffer ; (CBufferLoad)
+llvm.dx.resource.store.rawbuffer ; (RawBufferStore)
+llvm.dx.resource.store.typedbuffer ; (BufferStore)
+llvm.dx.resource.store.texture ; (TextureStore)
+llvm.dx.resource.store.texturesample ; (TextureStoreSample)
+llvm.dx.resource.writesamplerfeedback ; (WriteSamplerFeedback)
+llvm.dx.resource.writesamplerfeedbackbias ; (WriteSamplerFeedbackBias)
+llvm.dx.resource.writesamplerfeedbacklevel ; (WriteSamplerFeedbackLevel)
+llvm.dx.resource.writesamplerfeedbackgrad ; (WriteSamplerFeedbackGrad)
+llvm.dx.resource.updatecounter ; (BufferUpdateCounter)
+llvm.dx.resource.getdimensions ; (GetDimensions)
+llvm.dx.resource.sample ; (Sample)
+llvm.dx.resource.samplelevel ; (SampleLevel)
+llvm.dx.resource.samplebias ; (SampleBias)
+llvm.dx.resource.samplegrad ; (SampleGrad)
+llvm.dx.resource.samplecmp ; (SampleCmp)
+llvm.dx.resource.samplecmpbias ; (SampleCmpBias)
+llvm.dx.resource.samplecmpgrad ; (SampleCmpGrad)
+llvm.dx.resource.samplecmplevel ; (SampleCmpLevel)
+llvm.dx.resource.samplecmplevelzero ; (SampleCmpLevelZero)
+llvm.dx.resource.texturesamplepos ; (Texture2DMSGetSamplePosition)
+llvm.dx.resource.gather ; (TextureGather)
+llvm.dx.resource.gathercmp ; (TextureGatherCmp)
+llvm.dx.resource.gatherraw ; (TextureGatherRaw)
+llvm.dx.resource.atomicbinop ; (AtomicBinOp)
+llvm.dx.resource.atomiccmpxchg ; (AtomicCompareExchange)
```

Comment:
This and `pos` for position above seem to be the only places we use abbreviations of the dxil op name, maybe better to write it all out everywhere?

Reply:
I think these two abbreviations are both defensible:
- In the "pos" case we're abbreviating from "Texture2DMSGetSamplePosition" to "texturesamplepos". I'm fine with calling this "texturesampleposition" rather than "pos", but it still wouldn't match and I think saying "texture" rather than "texture2dms" is better here.
- "cmpxchg" is an extremely common abbreviation for "compare and exchange" and easy to parse at a glance, whereas "compareexchange" is arguably harder to read.

I don't feel super strongly here, but I have a weak preference for keeping these as is.

Reply:
Any reason to keep it as-is is good enough for me